### ISSUE TITLE
[TASK] Evaluate and try the feasibility to run tests in containered K8s cluster

### ISSUE THREAD

**Issue by innobead** [longhorn/longhorn#8158]
## What's the task? Please describe

https://github.com/longhorn/longhorn/issues/5693#issuecomment-1991119107

## Additional context

cc @longhorn/qa

**Comment by chriscchien** [comment]
If we use custom k3d images with iscsiadm binary installed([ref](https://github.com/chriscchien/k3s/blob/iscsi/iscsi_utils/Dockerfile.template)) to deploy Longhor, all Longhorn compoments can running without crash.

However longhorn volume can not attach success because of some iscsi service not really enabled in the container.
```
2024-03-14T02:41:15.233744069Z time="2024-03-14T02:41:15Z" level=warning msg="test-e-0: time=\"2024-03-14T02:39:08Z\" level=warning msg=\"Failed to discover\" func=\"iscsidev.(*Device).StartInitator\" file=\"iscsi.go:161\" error=\"failed to execute: /usr/bin/nsenter [nsenter --mount=/host/proc/2260/ns/mnt --net=/host/proc/2260/ns/net iscsiadm -m discovery -t sendtargets -p 10.42.0.13], output , stderr iscsiadm: Could not open /run/lock/iscsi: No such file or directory\\niscsiadm: Could not open /run/lock/iscsi: No such file or directory\\niscsiadm: Could not add new discovery record.\\n: exit status 6\"" func="controller.(*InstanceHandler).printInstanceLogs" file="instance_handler.go:467"
2024-03-14T02:41:15.233757508Z time="2024-03-14T02:41:15Z" level=warning msg="test-e-0: time=\"2024-03-14T02:39:08Z\" level=warning msg=\"Nodes cleaned up for iqn.2019-10.io.longhorn:test\" func=\"iscsidev.(*Device).StartInitator\" file=\"iscsi.go:168\"" func="controller.(*InstanceHandler).printInstanceLogs" file="instance_handler.go:467"
```

The k3d base image is very limited, it does not have package manager apk and even can not use modprobe
```
/ # apk
sh: apk: not found
/ # 
/ # modprobe iscsi_tcp
modprobe: can't change directory to '5.14.21-150500.55.44-default': No such file or directory
/ # 
/ # uname -r
5.14.21-150500.55.44-default
/ # 
/ # cat /etc/os-release 
PRETTY_NAME="K3s v1.29.2+k3s1"
```

For the [issue](https://github.com/longhorn/longhorn/issues/5693#issue-1649600025) reporter, I tihnk they built their k8s container by their own script here([ref1](https://github.com/naiming-zededa/eve/blob/master/pkg/kube/Dockerfile), [ref2](https://github.com/naiming-zededa/eve/blob/master/pkg/kube/cluster-init.sh)).

In conclusion, 
-  It's hard to congiure k3d container to have all necessary service Longhorn needed enable because the image is very [limited](https://k3d.io/v5.3.0/faq/faq/#longhorn-in-k3d).
-  I will keep working on this by study the issue reporter's build script after recent release, sorry for the lack of progress, thank you.

**Comment by innobead** [comment]
@chriscchien it's not about `iscsiadm` not found, but rather `Could not open /run/lock/iscsi: No such file`. Have you checked if `/run/lock` exists somehow? 

Even though the above issue is resolved, still not sure if `iscsi_tcp` kernel module can be used properly from the host into the container.
